turtle.py

```diff
@@ -0,0 +1,128 @@
+#submitted by Keren Or Hadad 208025205
+
+import string
+
+
+class PostOffice:
+
+    def __init__(self, usernames):
+        self.message_id = 0
+        self.read = False
+        self.boxes = {user: [] for user in usernames}
+
+    def send_message(self, sender, recipient, message_body, urgent=False):
+        """
+        Send a message to a recipient.
+
+        sender : str :The message sender's username.
+        recipient : str: The message recipient's username.
+        message_body : str: The body of the message.
+        urgent : bool, optional:The urgency of the message. Urgent messages appear first.
+        Returns: int: The message ID, auto incremented number.
+        KeyError: If the recipient does not exist.
+
+        """
+        user_box = self.boxes[recipient]
+        self.message_id = self.message_id + 1
+        message_details = {
+            'id': self.message_id,
+            'body': message_body,
+            'sender': sender,
+            'read': self.read
+        }
+        if urgent:
+            user_box.insert(0, message_details)
+        else:
+            user_box.append(message_details)
+        return self.message_id
+
+    def read_inbox(self, user, number_of_message=0):
+        """
+        The function return the messages of a user.
+
+        user : str: The user's name.
+        number_of_message : str: How mach message to return. If the user don't move any number, return all the message.
+        Returns: list: The message in g-mail of user.
+        KeyError: If the user does not exist.
+        """
+
+        len_message = len(self.boxes[user])
+        # Take only message that not read.
+        message_read = [message for message in self.boxes[user] if message['read'] is False]
+        # Take only message the user ask.
+        message_read = message_read[0:number_of_message] if number_of_message > 0 else message_read
+
+        for message in message_read:
+            for i in range(len_message):
+                if self.boxes[user][i]['id'] == message['id']:
+                    self.boxes[user][i]['read'] = True
+
+        return message_read
+
+    def search_inbox(self, user, part_message):
+        """
+        The function find messages (of the user) that containing the part message.
+
+        user : str: The user's name.
+        part_message : str: Sub message.
+        Returns: list: The messages that containing the part message.
+        KeyError: If the user does not exist.
+        """
+        return [message for message in self.boxes[user] if str(message['body']).__contains__(part_message)]
```

Comment:
Use in and not contains , methods with underscores should not be public.
https://stackoverflow.com/questions/3437059/does-python-have-a-string-contains-substring-method